### PR TITLE
feat: add dev preview feature

### LINKED ISSUE #72: Provide feature flags to enable specific plugins
Generally, feature flags allow us to activate or deactivate specific features of our application. 
In our case, we'll be allowing the SDK users to use feature flags to activate certain plugins.

This requires further investigation and discussion before starting out.

### PR DESCRIPTION
**Pull request by namit-chandwani**
Closes #72 
Based on discussion: https://github.com/aerogear/charmil/discussions/81

Still WIP.

## TODO
- [x] ~~Add inline documentation~~
- [x] ~~Add external docs~~
- [ ] Internationalize all strings
- [ ] Replace fmt.Println() statements with suitable logger functions
- [ ] Add tests
- [ ] Replace `.env` config storage approach with a better one, once issue: [#69] is resolved
---
## `config.env` file content:
```env
CHARMIL_DEV_PREVIEW_ENABLED="true" (or) "false"
```

### CODE REVIEW

File: examples/plugins/echo/locals/en/echo.en.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+echo.cmd.use:
+  description: "The one-line usage message"
+  one: "echo [string(s) to echo]"
+
+echo.cmd.short:
+  description: "The short description shown in the 'help' output"
+  one: "Prints given strings to stdout"
+
+echo.cmd.example:
```
Comment (wtrocki) [inline examples/plugins/echo/locals/en/echo.en.yaml:9]: why we need this

File: examples/plugins/echo/main.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	echoCmd := &cobra.Command{
+		Use:          f.Localizer.LocalizeByID("echo.cmd.use"),
+		Short:        f.Localizer.LocalizeByID("echo.cmd.short"),
+		Example:      f.Localizer.LocalizeByID("echo.cmd.example"),
+		SilenceUsage: true,
+		Args:         cobra.MinimumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			message := strings.Join(args, " ")
+
+			if uppercase {
```
Comment (wtrocki) [inline examples/plugins/echo/main.go:41]: why this logic?

File: pkg/preview/preview.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	"os"
+	"strconv"
+
+	"github.com/fatih/color"
+	"github.com/joho/godotenv"
+	"github.com/spf13/cobra"
+)
+
+type devPreview struct {
+	commands  []*cobra.Command
```
Comment (wtrocki) [inline pkg/preview/preview.go:14]: what are commands for

File: pkg/preview/preview.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	dp.isEnabled = boolVal
+}
+
+func UpdateFlagValue(cmd *cobra.Command, args []string) {
+	if cmd.Flags().Changed("dev_preview") {
+		env, err := godotenv.Unmarshal("CHARMIL_DEV_PREVIEW_ENABLED=" + strconv.FormatBool(dp.isEnabled))
```
Comment (wtrocki) [inline pkg/preview/preview.go:41]: Lets not write code we do not plan merging

**Comment by wtrocki** [comment]
@namit-chandwani this breaks contract we agreed on that we focus on usage and docunention frist

**Comment by namit-chandwani** [comment]
> @namit-chandwani twe agreed on that we focus on usage and docunention frist. Do you mind adding dev usage docs. it is hard to review this.

Sure, I'll add them and ping you once done

File: examples/plugins/echo/main.go (hunk 2 above)
Comment (namit-chandwani) [reply examples/plugins/echo/main.go:54]: That's just a part of an example plugin that I've added and doesn't do much.
I wanted to have a basic and minimal single command example plugin, so I chose to add this. 😅

All it does is clone the functionality of the Linux `echo` command ie. print all the strings passed as arguments, into stdout.
It has 3 local flags:
- **uppercase**: If this is enabled, strings are converted to uppercase before printing into stdout
- **lowercase**: If this is enabled, strings are converted to lowercase before printing into stdout
- **repeat**: The value that is passed into this flag is the number of times the input strings will be printed to stdout

File: pkg/preview/preview.go (hunk 3 above)
Comment (namit-chandwani) [reply pkg/preview/preview.go]: `commands` stores all the commands that have been added for dev preview by the host CLI developer.

I've added comments for the same in the file now, please have a look.

**Comment by wtrocki** [comment]
lets leave this PR for now. It is not mergable unti.we have config anyway

**Comment by namit-chandwani** [comment]
> lets leave this PR for now. It is not mergable unti.we have config anyway

Agreed. Will start working on the config issue now and come back to this PR only once that's done

File: examples/plugins/echo/main.go (hunk 2 above)
Comment (wtrocki) [reply examples/plugins/echo/main.go:54]: As we discussed I think it is best for examples to be minimal.
Less non functional code then less things that people need to read

File: pkg/preview/preview.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	for _, cmd := range dp.commands {
+		updateCmdFields(cmd)
+	}
+}
+
+// updateCmdFields modifies fields of the command passed in
+// as argument with appropriate values to mark it as preview.
+func updateCmdFields(cmd *cobra.Command) {
+	cmd.Annotations = map[string]string{"channel": "preview"}
+	cmd.Short = color.HiCyanString("[Preview] ") + cmd.Short
```
Comment (wtrocki) [inline pkg/preview/preview.go:83]: Devs would probably want this configurable

**Review by wtrocki — CHANGES_REQUESTED** [review]
I think you got this rhoas inspired code right but we need to keep it short and do less opiniated and non functional code.

**Comment by wtrocki** [comment]
I did remove this feature from rhoas as it wasn't really well done.

**Comment by namit-chandwani** [comment]
> Not sure how to drive that from now on.

If it's fine with you, I can try coming up with a different approach to achieve this.
As there's already a config management system in place now, there won't be any blockage to this PR like it was before

> I did remove this feature from rhoas as it wasn't really well done.

Can you please let me know what were the things that you found wrong with the Rhoas implementation?
I'll try to avoid those in my approach.

**Comment by wtrocki** [comment]
On top of pririoties I will leave that PR for now and focus on validation of config in the starter that were merged.

**Comment by namit-chandwani** [comment]
> On top of pririoties I will leave that PR for now and focus on validation of config in the starter that were merged.

Got it. Will switch my focus to that

**Comment by ankithans** [comment]
shall we close it @namit-chandwani ?

**Comment by namit-chandwani** [comment]
> shall we close it @namit-chandwani , while keeping the branch

Sure, closing it for now. 
Will open it back when needed

**Pull request closed without merging**